crates/c-api/src/config.rs

```diff
@@ -371,11 +389,10 @@ unsafe impl MemoryCreator for CHostMemoryCreator {
             maximum.unwrap_or(usize::MAX),
             reserved_size_in_bytes.unwrap_or(0),
             guard_size_in_bytes,
-            memory.as_mut_ptr(),
+            &mut memory,
         );
         match error {
             None => {
-                let memory = unsafe { memory.assume_init() };
```

Comment:
This is a much safer API, because we don't need to assume they will sent `env` and `finalizer` to nullptr if they don't need them.